### PR TITLE
[TS] LPS-177307 Performance issue with v4_3_4.DLFileEntryTypeDDMFieldAttributeUpgradeProcess

### PR DESCRIPTION
**Pull request by IstvanD**
Dear Team,

Could you please review this pull request?

Motivation
=======
When ~ 1000000 entries exist in the `DDMFieldAttribute` table, [DLFileEntryTypeDDMFieldAttributeUpgradeProcess](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_4/DLFileEntryTypeDDMFieldAttributeUpgradeProcess.java) will take a substantial amount of time to complete. This is because the SQL queries in the upgrade process don't use indexing, and perform full table scans instead.
When the `DDMField` and `DDMFieldAttribute` tables are created in [DDMFieldUpgradeProcess](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_1_0/DDMFieldUpgradeProcess.java), their indexes are not created along with them, leading to a suboptimal query time.

Proposed Solution
========
Adding the `DDMField` and `DDMFieldAttribute` table's indexes during the upgrade and restructuring the queries in `DLFileEntryTypeDDMFieldAttributeUpgradeProcess` to utilize the indexes of the `DDMStructureLink`, `DDMField` and `DDMFieldAttribute` tables (thanks to @fabian-bouche-liferay for the idea!).

Steps to Verify
========
Please see the linked LPS for the reproduction steps.
https://issues.liferay.com/browse/LPS-177307

Thank you and best regards,
István

### CODE REVIEW

**Comment by IstvanD** [comment]
Dear Team,

As the solution to the issue is urgent to the customer, would it be OK if I already build a hotfix?

Thank you and best regards,
István

**Comment by IstvanD** [comment]
Hi @carolmariaabb 

Thank you for the answer.

Best regards,
István

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_4/DLFileEntryTypeDDMFieldAttributeUpgradeProcess.java [hunk 1]
```diff
@@ -33,19 +36,31 @@
...
+	public DLFileEntryTypeDDMFieldAttributeUpgradeProcess(
+		CompanyLocalService companyLocalService) {
+
+		_companyLocalService = companyLocalService;
+	}
+
 	@Override
 	protected void doUpgrade() throws Exception {
+		_companyLocalService.forEachCompanyId(
+			companyId -> _upgradePerCompany(companyId));
```
Comment (carolmariaabb) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_4/DLFileEntryTypeDDMFieldAttributeUpgradeProcess.java:48]: @IstvanD can we rename `_upgradePerCompany` to `_updateDDMFieldAttribute`?

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Comment by carolmariaabb** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_4/DLFileEntryTypeDDMFieldAttributeUpgradeProcess.java [hunk 2]
```diff
@@ -59,7 +74,9 @@ protected void doUpgrade() throws Exception {
 					"update DDMFieldAttribute set smallAttributeValue = ? " +
 						"where fieldAttributeId = ? ");
 
-			preparedStatement1.setString(1, "numeric");
+			preparedStatement1.setLong(1, _DL_FILE_ENTRY_TYPE_CLASS_NAME_ID);
```
[the remark this answers is not in the document]
Comment (IstvanD) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v4_3_4/DLFileEntryTypeDDMFieldAttributeUpgradeProcess.java]: Hi @carolmariaabb I have force-pushed the changes.

**Review by carolmariaabb — APPROVED** [review]
LGTM!

**Comment by timpak** [comment]
ci:test:bundle

**Comment by epark18** [comment]
@IstvanD hey, I tested this locally, and it's still taking a very long time for the upgrade process. 
`2023-03-14 00:14:09.099 INFO  [main][UpgradeProcess:127] Upgrading com.liferay.dynamic.data.mapping.internal.upgrade.v4_1_0.DDMFieldUpgradeProcess in 7198045 ms` 
It's been stuck on `v4_3_4.DLFileEntryTypeDDMFieldAttributeUpgradeProcess` for a while as well.

**Comment by IstvanD** [comment]
Hi @epark18 it takes a long time, but it is much faster now.
Before the fix, v4_3_4.DLFileEntryTypeDDMFieldAttributeUpgradeProcess wasn't finished after a day run, while now (on my system a least). it took ~15 minutes to complete.

**Comment by IstvanD** [comment]
Dear Team,

As I see the tests passed.
Please let me know if the pull request is ready to be forwarded.

Thank you!

**Comment by epark18** [comment]
@IstvanD Yes, I actually tried the upgrade again and it was much quicker. 
`v4_3_4.DLFileEntryTypeDDMFieldAttributeUpgradeProcess in 663377 ms`
`v4_1_0.DDMFieldUpgradeProcess in228693 ms`
Looks good to me on QA side.
cc/ @timpak

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
ci:stop

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by carolmariaabb** [comment]
ci:forward

**Pull request closed without merging**